Sources/Basic/Lock.swift

```diff
@@ -25,3 +26,67 @@ public struct Lock {
         return try body()
     }
 }
+
+enum ProcessLockError: Swift.Error {
+    case unableToAquireLock(errno: Int32)
+}
+
+/// Provides functionality to aquire a lock on a file via POSIX's flock() method.
+/// It can be used for things like serializing concurrent mutations on a shared resource
+/// by mutiple instances of a process.
+public final class FileLock {
+    /// The name of the lock, used in filename of the lock file.
+    let name: String
+
+    /// The directory where the lock file should be created.
+    let cachePath: AbsolutePath
+
+    /// File descriptor to the lock file.
+    private var fd: Int32? = nil
+
+    /// Path to the lock file.
+    private var lockFile: AbsolutePath {
+        return cachePath.appending(component: name + ".lock")
+    }
+
+    /// Create an instance of process lock with a name and the path where
+    /// the lock file can be created.
+    ///
+    /// Note: The cache path should be a valid directory.
+    public init(name: String, cachePath: AbsolutePath) {
+        self.name = name
+        self.cachePath = cachePath
+    }
+
+    /// Try to aquire a lock. This method will block until lock the already aquired by other process.
+    ///
+    /// Note: This method can throw if underlying POSIX methods fail.
+    public func lock() throws {
+        // Open the lock file.
+        let fp = libc.fopen(lockFile.asString, "w")
+        if fp == nil {
+            throw FileSystemError(errno: errno)
+        }
+        // Save the fd to close and remove lock later.
+        fd = fileno(fp)
+        // Aquire lock on the file.
+        guard flock(fd!, LOCK_EX) == 0 else {
```

Comment:
This should have an `errno == EINTR` loop.

We may want to consider implementing which has the *semantics* of a file lock, when exclusively used by the SwiftPM class which creates it, but which doesn't actually use `flock`. The typical implementation here is to atomic write the process info to a file and then point lock file (as a symbolic link) to it, if that fails, then check the contents of the file to see if that PID is alive, and then wait for that file to be removed (along with a timeout/retry loop in case that process dies).

A couple refs:
- http://0pointer.de/blog/projects/locking.html
- http://llvm.org/docs/doxygen/html/LockFileManager_8cpp_source.html

Note that the first ref talks about how the approach I describe isn't reliable on places where you can't effectively do an atomic FS write (an NFS scenario), so this may night be the right approach. It is however the approach LLVM is currently taking, for better or worse.

This advantage of this approach is that it is more amenable to reuse of our other APIs (for example, the work we would need to do to support timeouts, which is something we should certainly have support for).

Disclaimer: I don't have much experience with actually using file system locking. I know enough to be paranoid. :)